src/smt/theory_datatype.cpp

```diff
@@ -389,10 +389,11 @@ namespace smt {
     final_check_status theory_datatype::final_check_eh() {
         int num_vars = get_num_vars();
         final_check_status r = FC_DONE;
+        final_check_st _guard(this); // RAII for managing state
         for (int v = 0; v < num_vars; v++) {
             if (v == static_cast<int>(m_find.find(v))) {
                 enode * node = get_enode(v);
-                if (occurs_check(node)) {
+                if (!oc_cycle_free(node) && occurs_check(node)) {
```

Comment:
we only need to occurs check expressions of a datatype sort. 
Both the old and new code postpone these filters until after a stack has been updated.

Reply:
If we start from variables of this theory, shouldn't they all be of a datatype sort? Or do these variables also cover recognizers?

Reply:
Correct, so my comment should be retracted